packages/@aws-cdk/aws-apigateway/lib/deployment.ts

```diff
@@ -170,20 +177,18 @@ class LatestDeploymentResource extends CfnDeployment {
    * Hooks into synthesis to calculate a logical ID that hashes all the components
    * add via `addToLogicalId`.
    */
-  public validate() {
+  protected prepare() {
     // if hash components were added to the deployment, we use them to calculate
     // a logical ID for the deployment resource.
     if (this.hashComponents.length === 0) {
       this.lazyLogicalId = this.originalLogicalId;
     } else {
       const md5 = crypto.createHash('md5');
       this.hashComponents
-        .map(c => cdk.resolve(c))
+        .map(c => this.node.resolve(c))
```

Comment:
That's interesting - /me likey.

Reply:
Are we prepared to have unresolved tokens in error messages?

https://github.com/awslabs/aws-cdk/blob/3a0d5b5cf98e3cbcbd6139ef1e07135b4bece5f7/packages/@aws-cdk/aws-kms/lib/key.ts#L71

https://github.com/awslabs/aws-cdk/blob/3a0d5b5cf98e3cbcbd6139ef1e07135b4bece5f7/packages/@aws-cdk/aws-lambda/lib/lambda-ref.ts#L419

Not entirely clear how this should work if it can't resolve:

https://github.com/awslabs/aws-cdk/blob/3a0d5b5cf98e3cbcbd6139ef1e07135b4bece5f7/packages/@aws-cdk/aws-s3/lib/util.ts#L35

The lazy arguments to `Token` will need to take the `SynthesisContext` as an argument as well so they can resolve or JSONify inner structures:

https://github.com/awslabs/aws-cdk/blob/3a0d5b5cf98e3cbcbd6139ef1e07135b4bece5f7/packages/@aws-cdk/aws-ecr/lib/repository.ts#L129

https://github.com/awslabs/aws-cdk/blob/3a0d5b5cf98e3cbcbd6139ef1e07135b4bece5f7/packages/@aws-cdk/aws-logs/lib/cross-account-destination.ts#L98

All still in favor?